File: jar/src/main/java/jline/api/cache/Cache_ttl_hlru.java
/**
 * @file TTL (characteristic-time) approximation for h-LRU / LRU(m) caches
 *
 * Steady-state list occupancy probabilities of the list-based h-LRU (LRU(m))
 * replacement policy under the characteristic-time approximation (Gast and
 * Van Houdt, SIGMETRICS 2015). For h=1 this reduces exactly to the Che
 * approximation for LRU.
 *
 * @since LINE 3.0
 */
package jline.api.cache;

import jline.util.matrix.Matrix;

public final class Cache_ttl_hlru {
    private Cache_ttl_hlru() {}

    /**
     * Steady-state list occupancy probabilities for an h-LRU cache.
     *
     * @param lambda per-class request rates; lambda[v] is (n x h+1) with the
     *               same per-item rate replicated across list slices (the MVA
     *               cache analyzer layout); rates are aggregated over classes
     * @param m      (1 x h) list capacities
     * @return (n x h+1) probabilities; column 0 = not cached, column 1+l = in
     *         list l
     */
    public static Matrix cache_ttl_hlru(Matrix[] lambda, Matrix m) {
        int n = lambda[0].getNumRows();
        int h = m.length();
        int rateCol = Math.min(1, lambda[0].getNumCols() - 1);
        Matrix lam = new Matrix(n, 1);
        for (int v = 0; v < lambda.length; v++) {
            for (int k = 0; k < n; k++) {
                lam.set(k, 0, lam.get(k, 0) + lambda[v].get(k, rateCol));
            }
        }

        Matrix t = Cache_t_hlru.cache_t_hlru(lam, m);
        double[] tv = new double[h];
        for (int l = 0; l < h; l++) {
            tv[l] = t.get(0, l);
        }

        Matrix pij = new Matrix(n, h + 1);
        for (int k = 0; k < n; k++) {
            double[] w = Cache_t_hlru.levelWeights(lam.get(k, 0), tv, h);
            double sum = 0.0;
            for (int s = 0; s <= h; s++) {
                sum += w[s];
            }
            for (int s = 0; s <= h; s++) {
                pij.set(k, s, w[s] / sum);
            }
        }
        return pij;
    }
}

File: jar/src/main/kotlin/jline/api/cache/Cache_ttl_hlru.kt
/**
 * @file TTL-based H-LRU Cache Analysis
 * 
 * Implements TTL approximation for Hierarchical LRU (H-LRU) cache systems.
 * Combines time-to-live eviction modeling with hierarchical LRU replacement
 * policies for multi-level cache performance analysis.
 * 
 * @since LINE 3.0
 */
package jline.api.cache

import jline.util.matrix.Matrix
import org.apache.commons.math3.util.FastMath

/**
 * Solve hierarchical least-recently-used caches h-LRU using the TTL approximation.
 *
 * @param lambda - Array of matrices representing request arrival rates from users to items of individual lists.
 * @param m      - Matrix representing cache capacity vector.
 * @return Matrix - A matrix containing the computed steady-state probabilities for the cache.
 */

fun cache_ttl_hlru(lambda: Array<Matrix>, m: Matrix): Matrix {
    val lambda2 = Matrix(lambda[0].numRows, lambda[0].numCols - 1)
    for (i in 0..<lambda[0].numRows) {
        for (j in 0..<lambda[0].numCols - 1) {
            lambda2[i, j] = lambda[0][i, j]
        }
    }


    val n = lambda2.numRows
    val h = m.numCols

    val t = cache_t_hlru(lambda2, m)

    val probh = Matrix(n, h) // steady state at list h
    val prob0 = Matrix(n, 1) // steady state at list 0
    val temp1 = Matrix(n, 1)
    temp1.ones()
    val temp2 = Matrix(n, 1)


    for (k in 0..<n) {
        for (s in 0..<h) {
            temp1[k, 0] = temp1[k, 0] * (1 - FastMath.exp(-lambda2[k, s] * t[s]))
        }

        var middtemp = 1.0
        var middtemp2 = 0.0
        for (l in 0..<h - 1) {
            for (s in 0..l) {
                middtemp = middtemp * (1 - FastMath.exp(-lambda2[k, s] * t[s]))
            }
            middtemp2 += middtemp
        }
        temp2[k, 0] = FastMath.exp(-lambda2[k, h - 1] * t[h - 1]) * (1 + middtemp2)

        probh[k, 0] = temp1[k, 0] / (temp1[k, 0] + temp2[k, 0])
        prob0[k, 0] = 1 / (temp1[k, 0] + temp2[k, 0]) / FastMath.exp(lambda2[k, h - 1] * t[0, h - 1])
    }


    return Matrix.concatColumns(prob0, probh, Matrix(n, h + 1))
}
/**
 * Cache ttl hlru algorithms
 */
@Suppress("unused")
class CacheTtlHlruAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}